src/lib.rs

```diff
@@ -0,0 +1,1235 @@
+    /// Get the index of the entry block.
+    fn entry_block(&self) -> Block;
+
+    /// Provide the range of instruction indices contained in each block.
+    fn block_insns(&self, block: Block) -> InstRange;
+
+    /// Get CFG successors for a given block.
+    fn block_succs(&self, block: Block) -> &[Block];
+
+    /// Get the CFG predecessors for a given block.
+    fn block_preds(&self, block: Block) -> &[Block];
+
+    /// Get the block parameters for a given block.
+    fn block_params(&self, block: Block) -> &[VReg];
+
+    /// Determine whether an instruction is a call instruction. This is used
+    /// only for splitting heuristics.
+    fn is_call(&self, insn: Inst) -> bool;
+
+    /// Determine whether an instruction is a return instruction.
+    fn is_ret(&self, insn: Inst) -> bool;
+
+    /// Determine whether an instruction is the end-of-block
+    /// branch. If so, its operands at the indices given by
+    /// `branch_blockparam_arg_offset()` below *must* be the block
+    /// parameters for each of its block's `block_succs` successor
+    /// blocks, in order.
+    fn is_branch(&self, insn: Inst) -> bool;
+
+    /// If `insn` is a branch at the end of `block`, returns the
+    /// operand index at which outgoing blockparam arguments are
+    /// found. Starting at this index, blockparam arguments for each
+    /// successor block's blockparams, in order, must be found.
+    ///
+    /// It is an error if `self.inst_operands(insn).len() -
+    /// self.branch_blockparam_arg_offset(insn)` is not exactly equal
+    /// to the sum of blockparam counts for all successor blocks.
+    fn branch_blockparam_arg_offset(&self, block: Block, insn: Inst) -> usize;
+
+    /// Determine whether an instruction is a safepoint and requires a stackmap.
+    ///
+    /// Strictly speaking, these two parts (is a safepoint, requires a
+    /// stackmap) are orthogonal. An instruction could want to see a
+    /// stackmap of refs on the stack (without forcing them), or it
+    /// could want all refs to be on the stack (without knowing where
+    /// they are). Only the latter strictly follows from "is a
+    /// safepoint". But in practice, both are true at the same time,
+    /// so we combine the two notions: for regalloc2, a "safepoint
+    /// instruction" is one that both forces refs onto the stack, and
+    /// provides a stackmap indicating where they are.
+    fn is_safepoint(&self, _: Inst) -> bool {
+        false
+    }
+
+    /// Determine whether an instruction is a move; if so, return the
+    /// Operands for (src, dst).
+    fn is_move(&self, insn: Inst) -> Option<(Operand, Operand)>;
+
+    // --------------------------
+    // Instruction register slots
+    // --------------------------
+
+    /// Get the Operands for an instruction.
+    fn inst_operands(&self, insn: Inst) -> &[Operand];
+
+    /// Get the clobbers for an instruction; these are the registers
+    /// that the instruction is known to overwrite, separate from its
```

Comment:
"known to overwrite" .. well, this is better, but I still think it fails to clearly
enough distinguish "overwrite" from "output operand".  Imagine a bizarre
add insn which *also* happens to (eg) write the constant 1337 to r15.  If the
instruction selector knows this fact and it needs to synthesise the constant
1337 right after this instruction, then it can simply reference r15.  So is r15 then
a clobber or an output operand?

I'd prefer if this was specified as "a clobbered register is one that will hold,
after the insn has executed, a value which is arbitrary and may be implementation
defined".  Hence pretty clearly expresses the fact that the contents of said
registers are no good for anything after the insn.  Or more precisely .. for said
registers, the only valid next-event is a write.

Reply:
Perhaps just say that a clobber is equivalent to a `FixedReg` `Def@Before` operand with a fresh, unique vreg. This means that conflicts between outputs operands and clobbers are invalid since the instruction is assigning two different vregs to the same preg.

Reply:
Updated, hopefully the latest description works; it starts more or less with @julian-seward1's "will hold a value which is arbitrary ..." definition and also mentions the equivalence to an early def of a fresh vreg with a fixed-reg constraint.